eclair-core/src/test/scala/fr/acinq/eclair/db/PgUtilsSpec.scala

```diff
@@ -0,0 +1,121 @@
+package fr.acinq.eclair.db
+
+import akka.actor.ActorSystem
+import com.opentable.db.postgres.embedded.EmbeddedPostgres
+import com.typesafe.config.{Config, ConfigFactory}
+import fr.acinq.eclair.TestUtils
+import fr.acinq.eclair.db.pg.PgUtils.JdbcUrlChanged
+import fr.acinq.eclair.db.pg.PgUtils.PgLock.{LockFailure, LockFailureHandler}
+import grizzled.slf4j.Logging
+import org.scalatest.funsuite.AnyFunSuite
+
+import java.io.File
+import java.util.UUID
+
+class PgUtilsSpec extends AnyFunSuite {
+
+  test("database lock") {
+    implicit val system: ActorSystem = ActorSystem()
+    val pg = EmbeddedPostgres.start()
+    val config = PgUtilsSpec.testConfig(pg.getPort)
+    val datadir = new File(TestUtils.BUILD_DIRECTORY, s"pg_test_${UUID.randomUUID()}")
+    datadir.mkdirs()
+    val instanceId1 = UUID.randomUUID()
+    // this will lock the database for this instance id
+    val db = Databases.postgres(config, instanceId1, datadir, LockFailureHandler.logAndThrow)
+
+    assert(
+      intercept[LockFailureHandler.LockException] {
+        // this will fail because the database is already locked for a different instance id
+        Databases.postgres(config, UUID.randomUUID(), datadir, LockFailureHandler.logAndThrow)
+      }.lockFailure === LockFailure.AlreadyLocked(instanceId1))
+
+    // we can renew the lease at will
+    db.obtainExclusiveLock()
+
+    // we wait longer than the lease interval, and make sure that the lock is still there
+    Thread.sleep(20_000)
+    assert(
+      intercept[LockFailureHandler.LockException] {
+        // this will fail because the database is already locked for a different instance id
+        Databases.postgres(config, UUID.randomUUID(), datadir, LockFailureHandler.logAndThrow)
+      }.lockFailure === LockFailure.AlreadyLocked(instanceId1))
+
+    // we close the first connection
+    db.dataSource.close()
+    while (!db.dataSource.isClosed) {
+      Thread.sleep(1000)
+    }
```

Comment:
Looks like you should rather extend `TestKitBase` and use `awaitCond`? Especially since you need an `ActorSystem`.

Reply:
Done with 8a844fa.

I didn't find a better way than using `Thread.sleep`. It seems there is no standard non-blocking way of waiting in tests. I looked into scalatest's `Waiters` but it's for a different purpose. `awaitCond` and `eventually` are for cases where we expect a statement to eventually return a certain value. But here we just want to wait.

Reply:
But this should work:

```
awaitCond(db.dataSource.isClosed)
```

Doesn't it? And it can be used in multiple places in this file.

I agree that you can't really replace the `Thread.sleep` that are waiting for the lease to expire on the postgres side but these ones can be improved.

Reply:
Totally! Done with f243a119d.